Sources/ProjectSpec/Target.swift

```diff
@@ -139,9 +139,26 @@ extension Target {
 
         let targetTemplatesDictionary: [String: JSONDictionary] = jsonDictionary["targetTemplates"] as? [String: JSONDictionary] ?? [:]
 
-        for (targetName, var target) in targetsDictionary {
+        // Recursively collects all nested template names of a given dictionary
+        func collectTemplates(of jsonDictionary: JSONDictionary, into allTemplates: inout [String], at index: Int = 0) {
+            guard let templates = jsonDictionary["templates"] as? [String] else {
+                return
+            }
+            var index = index
+            for template in templates where !allTemplates.contains(template) {
+                guard let templateDictionary = targetTemplatesDictionary[template] else {
+                    continue
+                }
+                allTemplates.insert(template, at: index)
+                collectTemplates(of: templateDictionary, into: &allTemplates, at: index)
+                index += 1
```

Comment:
From reading the code it's not immediately obvious to me what the index does?

Reply:
The index makes sure that templates are inserted in the right order, e.g. this template tree:
 - A
   - B
   - C
 - D
   - E

will result in the following template list: B,C,A,E,D
After collecting the list templates will be processed as previously.
This ensures that the leaf templates are always processed before processing the parent templates so that parent templates can override values.

I tried to make that more clear from the code by renaming the parameter. And now that you point it out I realized that the logic was wrong. I fixed that and added a test case that covers this specific example.